src/appleseed.python/bindblenderexport.cpp

```diff
@@ -84,17 +92,131 @@ namespace
 
     struct MLoopUV
     {
-        float uv[2];
-        int flag;
+        float   uv[2];
+        int     flag;
     };
 
     struct MPoly
     {
-        int loopstart;
-        int totloop;
-        short mat_nr;
-        char flag, pad;
+        int     loopstart;
+        int     totloop;
+        short   mat_nr;
+        char    flag, pad;
+    };
+
+
+    //
+    // This class stores and manipulates an arbitrary number of appleseed transform sequences
+    // from Blender particle and dupli instances.  Storing them in a C++
+    // std::vector allows them to be processed and inserted into the appleseed scene
+    // without the overhead of Python.
+    //
+    
+    class BlTransformLibrary
+    {
+      public:
+        BlTransformLibrary() = default;
+
+        ~BlTransformLibrary() = default;
+
+        void release()
```

Comment:
You should remove this and use `std::unique_ptr<>` instead of `foundation::auto_release_ptr<>`.

Reply:
Tried modifying to:

std::unique_ptr<BlTransformLibrary> create_bl_transform_library()
    {
        return std::make_unique<BlTransformLibrary>();
    }

Then the Boost.Python class definition to:
bpy::class_<BlTransformLibrary, std::unique_ptr<BlTransformLibrary>, boost::noncopyable>("BlTransformLibrary", bpy::no_init)

Compile failed with:
1>------ Build started: Project: appleseed.shaders, Configuration: Release x64 ------
2>------ Build started: Project: appleseed.python3, Configuration: Release x64 ------
1>Compiling OSL shaders...
2>bindblenderexport.cpp
2>D:\boost_dir\boost_1_69_0\boost/python/converter/as_to_python_function.hpp(40): error C2280: 'std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<_Ty>>::unique_ptr(const std::unique_ptr<_Ty,std::default_delete<_Ty>> &)': attempting to reference a deleted function
2>        with
2>        [
2>            _Ty=`anonymous-namespace'::BlTransformLibrary
2>        ]
2>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.16.27023\include\memory(2337): note: see declaration of 'std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<_Ty>>::unique_ptr'
2>        with
2>        [
2>            _Ty=`anonymous-namespace'::BlTransformLibrary
2>        ]
2>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.16.27023\include\memory(2337): note: 'std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<_Ty>>::unique_ptr(const std::unique_ptr<_Ty,std::default_delete<_Ty>> &)': function was explicitly deleted
2>        with
2>        [
2>            _Ty=`anonymous-namespace'::BlTransformLibrary
2>        ]
2>D:\boost_dir\boost_1_69_0\boost/python/converter/as_to_python_function.hpp(26): note: while compiling class template member function 'PyObject *boost::python::converter::as_to_python_function<T,Conversion>::convert(const void *)'
2>        with
2>        [
2>            T=std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,
2>            Conversion=boost::python::objects::class_value_wrapper<std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,boost::python::objects::make_ptr_instance<`anonymous-namespace'::BlTransformLibrary,boost::python::objects::pointer_holder<std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,`anonymous-namespace'::BlTransformLibrary>>>
2>        ]
2>D:\boost_dir\boost_1_69_0\boost/python/to_python_converter.hpp(82): note: see reference to function template instantiation 'PyObject *boost::python::converter::as_to_python_function<T,Conversion>::convert(const void *)' being compiled
2>        with
2>        [
2>            T=std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,
2>            Conversion=boost::python::objects::class_value_wrapper<std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,boost::python::objects::make_ptr_instance<`anonymous-namespace'::BlTransformLibrary,boost::python::objects::pointer_holder<std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,`anonymous-namespace'::BlTransformLibrary>>>
2>        ]
2>D:\boost_dir\boost_1_69_0\boost/python/to_python_converter.hpp(83): note: see reference to class template instantiation 'boost::python::converter::as_to_python_function<T,Conversion>' being compiled
2>        with
2>        [
2>            T=std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,
2>            Conversion=boost::python::objects::class_value_wrapper<std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,boost::python::objects::make_ptr_instance<`anonymous-namespace'::BlTransformLibrary,boost::python::objects::pointer_holder<std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,`anonymous-namespace'::BlTransformLibrary>>>
2>        ]
2>D:\boost_dir\boost_1_69_0\boost/python/to_python_converter.hpp(77): note: while compiling class template member function 'boost::python::to_python_converter<Src,boost::python::objects::class_value_wrapper<Src,boost::python::objects::make_ptr_instance<T2,boost::python::objects::pointer_holder<std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<_Ty>>,`anonymous-namespace'::BlTransformLibrary>>>,true>::to_python_converter(void)'
2>        with
2>        [
2>            Src=std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,
2>            T2=`anonymous-namespace'::BlTransformLibrary,
2>            _Ty=`anonymous-namespace'::BlTransformLibrary
2>        ]
2>D:\appleseed\appleseed_py37\src\appleseed.python\bindblenderexport.cpp(467): note: see reference to function template instantiation 'boost::python::to_python_converter<Src,boost::python::objects::class_value_wrapper<Src,boost::python::objects::make_ptr_instance<T2,boost::python::objects::pointer_holder<std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<_Ty>>,`anonymous-namespace'::BlTransformLibrary>>>,true>::to_python_converter(void)' being compiled
2>        with
2>        [
2>            Src=std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,
2>            T2=`anonymous-namespace'::BlTransformLibrary,
2>            _Ty=`anonymous-namespace'::BlTransformLibrary
2>        ]
2>D:\boost_dir\boost_1_69_0\boost/python/object/class_wrapper.hpp(39): note: see reference to class template instantiation 'boost::python::to_python_converter<Src,boost::python::objects::class_value_wrapper<Src,boost::python::objects::make_ptr_instance<T2,boost::python::objects::pointer_holder<std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<_Ty>>,`anonymous-namespace'::BlTransformLibrary>>>,true>' being compiled
2>        with
2>        [
2>            Src=std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,
2>            T2=`anonymous-namespace'::BlTransformLibrary,
2>            _Ty=`anonymous-namespace'::BlTransformLibrary
2>        ]
2>D:\boost_dir\boost_1_69_0\boost/python/object/class_metadata.hpp(254): note: see reference to class template instantiation 'boost::python::objects::class_value_wrapper<std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<_Ty>>,boost::python::objects::make_ptr_instance<T2,boost::python::objects::pointer_holder<std::unique_ptr<_Ty,std::default_delete<_Ty>>,`anonymous-namespace'::BlTransformLibrary>>>' being compiled
2>        with
2>        [
2>            _Ty=`anonymous-namespace'::BlTransformLibrary,
2>            T2=`anonymous-namespace'::BlTransformLibrary
2>        ]
2>D:\boost_dir\boost_1_69_0\boost/python/object/class_metadata.hpp(230): note: see reference to function template instantiation 'void boost::python::objects::class_metadata<W,X1,X2,X3>::maybe_register_pointer_to_python<T2>(T2 *,boost::mpl::false_ *,boost::mpl::false_ *)' being compiled
2>        with
2>        [
2>            W=`anonymous-namespace'::BlTransformLibrary,
2>            X1=std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,
2>            X2=boost::noncopyable,
2>            X3=boost::python::detail::not_specified,
2>            T2=`anonymous-namespace'::BlTransformLibrary
2>        ]
2>D:\boost_dir\boost_1_69_0\boost/python/object/class_metadata.hpp(229): note: see reference to function template instantiation 'void boost::python::objects::class_metadata<W,X1,X2,X3>::maybe_register_pointer_to_python<T2>(T2 *,boost::mpl::false_ *,boost::mpl::false_ *)' being compiled
2>        with
2>        [
2>            W=`anonymous-namespace'::BlTransformLibrary,
2>            X1=std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,
2>            X2=boost::noncopyable,
2>            X3=boost::python::detail::not_specified,
2>            T2=`anonymous-namespace'::BlTransformLibrary
2>        ]
2>D:\boost_dir\boost_1_69_0\boost/python/object/class_metadata.hpp(218): note: see reference to function template instantiation 'void boost::python::objects::class_metadata<W,X1,X2,X3>::register_aux2<T,use_callback>(T2 *,Callback)' being compiled
2>        with
2>        [
2>            W=`anonymous-namespace'::BlTransformLibrary,
2>            X1=std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,
2>            X2=boost::noncopyable,
2>            X3=boost::python::detail::not_specified,
2>            T=`anonymous-namespace'::BlTransformLibrary,
2>            T2=`anonymous-namespace'::BlTransformLibrary,
2>            Callback=use_callback
2>        ]
2>D:\boost_dir\boost_1_69_0\boost/python/object/class_metadata.hpp(218): note: see reference to function template instantiation 'void boost::python::objects::class_metadata<W,X1,X2,X3>::register_aux2<T,use_callback>(T2 *,Callback)' being compiled
2>        with
2>        [
2>            W=`anonymous-namespace'::BlTransformLibrary,
2>            X1=std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,
2>            X2=boost::noncopyable,
2>            X3=boost::python::detail::not_specified,
2>            T=`anonymous-namespace'::BlTransformLibrary,
2>            T2=`anonymous-namespace'::BlTransformLibrary,
2>            Callback=use_callback
2>        ]
2>D:\boost_dir\boost_1_69_0\boost/python/object/class_metadata.hpp(216): note: while compiling class template member function 'void boost::python::objects::class_metadata<W,X1,X2,X3>::register_aux(void *)'
2>        with
2>        [
2>            W=`anonymous-namespace'::BlTransformLibrary,
2>            X1=std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,
2>            X2=boost::noncopyable,
2>            X3=boost::python::detail::not_specified
2>        ]
2>D:\boost_dir\boost_1_69_0\boost/python/object/class_metadata.hpp(204): note: see reference to function template instantiation 'void boost::python::objects::class_metadata<W,X1,X2,X3>::register_aux(void *)' being compiled
2>        with
2>        [
2>            W=`anonymous-namespace'::BlTransformLibrary,
2>            X1=std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,
2>            X2=boost::noncopyable,
2>            X3=boost::python::detail::not_specified
2>        ]
2>D:\boost_dir\boost_1_69_0\boost/python/class.hpp(170): note: see reference to class template instantiation 'boost::python::objects::class_metadata<W,X1,X2,X3>' being compiled
2>        with
2>        [
2>            W=`anonymous-namespace'::BlTransformLibrary,
2>            X1=std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<`anonymous-namespace'::BlTransformLibrary>>,
2>            X2=boost::noncopyable,
2>            X3=boost::python::detail::not_specified
2>        ]
2>D:\boost_dir\boost_1_69_0\boost/python/class.hpp(589): note: see reference to class template instantiation 'boost::python::class_<`anonymous-namespace'::BlTransformLibrary,std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<_Ty>>,boost::noncopyable,boost::python::detail::not_specified>::id_vector' being compiled
2>        with
2>        [
2>            _Ty=`anonymous-namespace'::BlTransformLibrary
2>        ]
2>D:\boost_dir\boost_1_69_0\boost/python/class.hpp(589): note: while compiling class template member function 'boost::python::class_<`anonymous-namespace'::BlTransformLibrary,std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<_Ty>>,boost::noncopyable,boost::python::detail::not_specified>::class_(const char *,boost::python::no_init_t)'
2>        with
2>        [
2>            _Ty=`anonymous-namespace'::BlTransformLibrary
2>        ]
2>D:\appleseed\appleseed_py37\src\appleseed.python\bindblenderexport.cpp(460): note: see reference to function template instantiation 'boost::python::class_<`anonymous-namespace'::BlTransformLibrary,std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<_Ty>>,boost::noncopyable,boost::python::detail::not_specified>::class_(const char *,boost::python::no_init_t)' being compiled
2>        with
2>        [
2>            _Ty=`anonymous-namespace'::BlTransformLibrary
2>        ]
2>D:\appleseed\appleseed_py37\src\appleseed.python\bindblenderexport.cpp(452): note: see reference to class template instantiation 'boost::python::class_<`anonymous-namespace'::BlTransformLibrary,std::unique_ptr<`anonymous-namespace'::BlTransformLibrary,std::default_delete<_Ty>>,boost::noncopyable,boost::python::detail::not_specified>' being compiled
2>        with
2>        [
2>            _Ty=`anonymous-namespace'::BlTransformLibrary
2>        ]
2>D:\boost_dir\boost_1_69_0\boost/bind/placeholders.hpp(54): note: see reference to class template instantiation 'boost::arg<9>' being compiled
2>D:\boost_dir\boost_1_69_0\boost/bind/placeholders.hpp(53): note: see reference to class template instantiation 'boost::arg<8>' being compiled
2>D:\boost_dir\boost_1_69_0\boost/bind/placeholders.hpp(52): note: see reference to class template instantiation 'boost::arg<7>' being compiled
2>D:\boost_dir\boost_1_69_0\boost/bind/placeholders.hpp(51): note: see reference to class template instantiation 'boost::arg<6>' being compiled
2>D:\boost_dir\boost_1_69_0\boost/bind/placeholders.hpp(50): note: see reference to class template instantiation 'boost::arg<5>' being compiled
2>D:\boost_dir\boost_1_69_0\boost/bind/placeholders.hpp(49): note: see reference to class template instantiation 'boost::arg<4>' being compiled
2>D:\boost_dir\boost_1_69_0\boost/bind/placeholders.hpp(48): note: see reference to class template instantiation 'boost::arg<3>' being compiled
2>D:\boost_dir\boost_1_69_0\boost/bind/placeholders.hpp(47): note: see reference to class template instantiation 'boost::arg<2>' being compiled
2>D:\boost_dir\boost_1_69_0\boost/bind/placeholders.hpp(46): note: see reference to class template instantiation 'boost::arg<1>' being compiled
2>Done building project "appleseed.python3.vcxproj" -- FAILED.
3>------ Skipped Build: Project: INSTALL, Configuration: Release x64 ------
3>Project not selected to build for this solution configuration 
========== Build: 1 succeeded, 1 failed, 10 up-to-date, 1 skipped ==========

Not sure what I'm doing wrong.

Reply:
We'll look into this separately.